my-webxr-app/src/components/PCA.tsx

```diff
@@ -0,0 +1,175 @@
+import { Matrix, EVD } from 'ml-matrix';
+
+/** 
+ * Inspired by: https://medium.com/analytics-vidhya/understanding-principle-component-analysis-pca-step-by-step-e7a4bb4031d9 
+ * 
+ * Steps Involved in the PCA
+ *    1: Standardize the dataset.
+ *    2: Calculate the covariance matrix for the features in the dataset.
+ *    3: Calculate the eigenvalues and eigenvectors for the covariance matrix and sort eigenvalues and their corresponding eigenvectors.
+ *    4: Pick k eigenvalues and form a matrix of eigenvectors. Then Transform the original matrix.
+ * 
+ * Note: assertions in some functions could be turn on/off and passed to inner function to reduce computations.
+*/
+
+type MaybeMatrix = number[][] | Matrix;
+
+/**
+ * Converts the input dataset into a Matrix.
+ * 
+ * If the dataset is already a Matrix, it is returned as is. If the dataset is a 2D array of numbers, it is converted into a Matrix.
+ * 
+ * @param dataset - The dataset to convert into a Matrix. This can be a Matrix or a 2D array of numbers.
+ * @param assert - Optional. Whether to enable assertions. Default is true.
+ * @returns The dataset as a Matrix.
+ * @throws {Error} If the input dataset is an array but not a 2D array.
+ */
+export function convertToMatrix(dataset: MaybeMatrix, assert: boolean = true): Matrix {
+    let datasetMatrix: Matrix;
+    if (Array.isArray(dataset)) {
+        // Lazy evaluation, assert is false the rest of condition will not be evaluated, reducing computation.
+        if (assert && !dataset.every(subArray => Array.isArray(subArray))) {
+            throw new Error("The input dataset must be a 2D array of numbers.");
+        }
+        datasetMatrix = new Matrix(dataset);
+    } else {
+        datasetMatrix = dataset;
+    } 
+    return datasetMatrix;
+}
+
+/**
+ * Standardizes the given dataset.
+ * 
+ * This is done by subtracting the mean and dividing by the standard deviation for each feature independently.
+ * 
+ * @param dataSetMatrix - The dataset to standardize, represented as a Matrix.
+ * @param assert - Optional. Whether to check if the standard deviation is zero. Default is true.
+ * @returns The standardized matrix of dataset.
+ * @throws {Error} If the standard deviation of a column in the dataset is zero and assert is true.
+ */
+export function standardizeDataset(dataSetMatrix: Matrix, assert: boolean = true): Matrix {
+    const mean = dataSetMatrix.mean('column');
+    const stdevs = dataSetMatrix.standardDeviation('column', { mean });
+
+    if (assert) {
```

Comment:
it is the constraint of the math, to prevent dividing by zero, ((data -mean)/std) on line 64.

Reply:
Why do we have this as an option? If the math breaks if the standard deviation is 0, why don't we just always check?

Reply:
There are two things that lead to this:
1.  It is quite rare to have a standard deviation of zero since that means we would have a data set of identical values.
2.  Repeating check on large datasets could be costly, so I want an option to make trade-off

Reply:
If it's something that will kill the program, we need to check for it always. Assertions are meant to check for logical assumptions, not to validate data. If we don't want to run this when the standard deviation is 0, we either need to check in the function, or never call it if we find the standard deviation to be 0 earlier.

Reply:
tbh, i am not sure how to deal with this when the std actually is zero, this ruin everything in this method. May look into this more and get back at you.

And you are absolutely right, this should kill the program.

Reply:
We do not want it to kill the program. The last thing we want to happen is for a user to upload a funky file and it crash the server.

Reply:
For this, I would just ignore the column where the std is zero and keep moving on (a for loop, an if check and continue statement). This will prevent the program from halting. 

STD=0  =>  Values are identical  =>  no variance in that column  =>  no meaningful PCA  => why even bother.

Working on the fix now, thank you for pointing it out. I need to learn how to use assertion.